json-rpc/src/runtime.rs

```diff
@@ -161,19 +213,37 @@ async fn rpc_endpoint_without_metrics(
                     ledger_info.ledger_info().timestamp_usecs(),
                 );
                 set_response_error(&mut resp, err);
+                log_response(trace_id, &resp, true);
 
                 warp::reply::json(&resp)
             }
         }
     } else {
         // single API call
-        let resp = rpc_request_handler(data, service, registry, ledger_info, LABEL_SINGLE).await;
+        let resp =
+            rpc_request_handler(data, service, registry, ledger_info, LABEL_SINGLE, trace_id).await;
+        log_response(trace_id, &resp, false);
+
         warp::reply::json(&resp)
     });
 
     Ok(Box::new(resp) as Box<dyn warp::Reply>)
 }
 
+fn log_response(trace_id: u64, resp: &JsonRpcResponse, is_batch: bool) {
+    let log = RpcResponseLog {
+        trace_id,
+        is_batch,
+        response_error: resp.error.is_some(),
+        response: resp,
+    };
+    if is_internal_error(resp.error.clone()) {
+        error!(log);
+    } else {
+        debug!(log);
+    }
+}
+
```

Comment:
I highly suggest avoiding this anit-pattern and use the logging macros directly where you want to log. This will ensure that you can get the proper file/line number/module included in the logs themselves.

Reply:
Yeah, in this case, it'll tell you whether it was in the batch / single API call  or error etc.

I think this might be one of the more acceptable spots of this anti-pattern, because it's highly generalized.

Reply:
I still suggest changing if possible, or write a short macro that just expands to this instead.

Reply:
suggestions are welcome, but I need more hint about what you mean, I'm trying my best with this doc: https://developers.libra.org/docs/rustdocs/libra_logger/index.html

Reply:
I'm guessing you are asking me log where the error created. 
this log is not for that purpose, this is at endpoint level to understand an internal unexpected error maybe returned, and we need to know what is that error.
For the case where error created, it's up to internal component to decide whether we need log, because some errors maybe handled at high level, for example, let's say account not found in database, if it is considered as an error at db level, but at get_account method, we return None instead of return error. 
Hence technically at the place we create the error, we don't know whether we returned the error to client as internal error.

Ideally, we should not have any error log here, but I'm sure we will see some in production.

Reply:
I'm asking that you remove and stop using a "log" function and instead just use the logging macros ("info", "error", etc) in-place of using a function like this.